File: src/utils/data_processor.py
"""
Data Processing Utilities
Handles cleaning, formatting, and standardizing scraped apartment data
"""

import re
from typing import List, Optional
from datetime import datetime, date
from ..types.types import FloorPlan, PropertySummary, DailyReport, BedroomDistribution, PriceRange


class DataProcessor:
    """
    Utility class for processing and cleaning apartment data
    """
    
    @staticmethod
    def parse_price(price_string: str) -> float:
        """Parse and clean price strings into numbers"""
        if not price_string:
            return 0.0
        
        # Handle ranges like "1,583 - 1,650" first
        if '-' in price_string:
            parts = price_string.split('-')
            prices = []
            for part in parts:
                # Extract digits and commas from each part
                clean_part = re.sub(r'[^0-9,.]', '', part.strip())
                try:
                    price = float(clean_part.replace(',', ''))
                    if price > 0:
                        prices.append(price)
                except ValueError:
                    continue
            return min(prices) if prices else 0.0
        
        # Handle "From $1,583" cases - find all numbers with commas
        matches = re.findall(r'\d+(?:,\d+)*(?:\.\d+)?', price_string)
        if matches:
            try:
                # Take the first number found
                return float(matches[0].replace(',', ''))
            except ValueError:
                pass
        
        return 0.0

    @staticmethod
    def parse_square_footage(sqft_string: str) -> int:
        """Parse square footage from strings"""
        if not sqft_string:
            return 0
        
        matches = re.search(r'(\d+(?:,\d+)*)', sqft_string)
        if matches:
            try:
                return int(matches.group(1).replace(',', ''))
            except ValueError:
                pass
        
        return 0

    @staticmethod
    def parse_bedroom_count(bedroom_string: str) -> int:
        """Parse bedroom count from strings"""
        if not bedroom_string:
            return 0
        
        bedroom_lower = bedroom_string.lower().strip()
        
        # Handle studio apartments
        if 'studio' in bedroom_lower:
            return 0
        
        # Extract number from strings like "1 Bed", "2 Bedroom", etc.
        matches = re.search(r'(\d+)', bedroom_lower)
        if matches:
            try:
                return int(matches.group(1))
            except ValueError:
                pass
        
        return 0

    @staticmethod
    def parse_bathroom_count(bathroom_string: str) -> float:
        """Parse bathroom count from strings, supports half bathrooms"""
        if not bathroom_string:
            return 0.0
        
        bathroom_lower = bathroom_string.lower().strip()
        
        # Handle half bathrooms
        if '1.5' in bathroom_lower or 'one and a half' in bathroom_lower:
            return 1.5
        elif '2.5' in bathroom_lower or 'two and a half' in bathroom_lower:
            return 2.5
        elif '3.5' in bathroom_lower or 'three and a half' in bathroom_lower:
            return 3.5
        
        # Extract number from strings like "1 Bath", "2 Bathroom", etc.
        matches = re.search(r'(\d+(?:\.\d+)?)', bathroom_lower)
        if matches:
            try:
                return float(matches.group(1))
            except ValueError:
                pass
        
        return 1.0  # Default to 1 bathroom if unclear

    @staticmethod
    def calculate_price_per_sqft(price: float, square_footage: int) -> float:
        """Calculate price per square foot"""
        if square_footage <= 0:
            return 0.0
        return round(price / square_footage, 2)

    @staticmethod
    def clean_amenities(amenities_raw: List[str]) -> List[str]:
        """Clean and standardize amenity names"""
        cleaned = []
        for amenity in amenities_raw:
            if not amenity:
                continue
            
            # Remove extra whitespace and convert to title case
            cleaned_amenity = ' '.join(amenity.strip().split())
            
            # Skip very short or very long amenities (likely not useful)
            if 2 <= len(cleaned_amenity) <= 50:
                cleaned.append(cleaned_amenity)
        
        # Remove duplicates while preserving order
        seen = set()
        result = []
        for amenity in cleaned:
            if amenity.lower() not in seen:
                seen.add(amenity.lower())
                result.append(amenity)
        
        return result

    @staticmethod
    def standardize_availability(availability_string: str) -> str:
        """Standardize availability strings"""
        if not availability_string:
            return "Unknown"
        
        availability_lower = availability_string.lower().strip()
        
        if 'available now' in availability_lower or 'immediate' in availability_lower:
            return "Available Now"
        elif 'coming soon' in availability_lower or 'waitlist' in availability_lower:
            return "Coming Soon"
        elif 'not available' in availability_lower or 'unavailable' in availability_lower:
            return "Not Available"
        else:
            # Try to extract and format date
            date_match = re.search(r'(\d{1,2}[-/]\d{1,2}[-/]\d{2,4})', availability_string)
            if date_match:
                return f"Available {date_match.group(1)}"
            
            return availability_string.strip()

    @staticmethod
    def validate_floor_plan(floor_plan: dict) -> bool:
        """Validate that a floor plan has minimum required data"""
        required_fields = ['name', 'price', 'property_name']
        
        for field in required_fields:
            if field not in floor_plan or not floor_plan[field]:
                return False
        
        # Price should be reasonable (between $500 and $10,000)
        price = floor_plan.get('price', 0)
        if not (500 <= price <= 10000):
            return False
        
        return True

    @staticmethod
    def generate_property_summary(floor_plans: List[FloorPlan]) -> PropertySummary:
        """Generate summary statistics for a property"""
        if not floor_plans:
            return PropertySummary(
                property_name="Unknown",
                total_floor_plans=0,
                price_range=PriceRange(min=0, max=0),
                avg_price_per_sq_ft=0.0,
                bedroom_distribution=BedroomDistribution(),
                available_units=0
            )
        
        property_name = floor_plans[0].property_name
        prices = [fp.price for fp in floor_plans if fp.price > 0]
        price_per_sqft = [fp.price_per_sq_ft for fp in floor_plans if fp.price_per_sq_ft > 0]
        
        # Count bedroom distribution
        bedroom_dist = BedroomDistribution()
        available_count = 0
        
        for fp in floor_plans:
            if fp.availability.lower() not in ['not available', 'unavailable']:
                available_count += 1
            
            if fp.bedrooms == 0:
                bedroom_dist.studio += 1
            elif fp.bedrooms == 1:
                bedroom_dist.one_bed += 1
            elif fp.bedrooms == 2:
                bedroom_dist.two_bed += 1
            elif fp.bedrooms == 3:
                bedroom_dist.three_bed += 1
            else:
                bedroom_dist.four_plus_bed += 1
        
        return PropertySummary(
            property_name=property_name,
            total_floor_plans=len(floor_plans),
            price_range=PriceRange(
                min=min(prices) if prices else 0,
                max=max(prices) if prices else 0
            ),
            avg_price_per_sq_ft=round(sum(price_per_sqft) / len(price_per_sqft), 2) if price_per_sqft else 0.0,
            bedroom_distribution=bedroom_dist,
            available_units=available_count
        )

    @staticmethod
    def format_timestamp() -> str:
        """Generate formatted timestamp string"""
        return datetime.now().isoformat()

    @staticmethod
    def format_date() -> str:
        """Generate formatted date string"""
        return date.today().isoformat()

File: src/utils/dataProcessor.ts
/**
 * Data Processing Utilities
 * Handles cleaning, formatting, and standardizing scraped apartment data
 */

import { FloorPlan, PropertySummary, DailyReport } from '../types/types';

export class DataProcessor {
  
  /**
   * Parse and clean price strings into numbers
   */
  static parsePrice(priceString: string): number {
    if (!priceString) return 0;
    
    // Remove everything except digits, commas, and periods
    const cleanPrice = priceString.replace(/[^0-9,.]/g, '');
    
    // Handle ranges like "1,583 - 1,650"
    if (cleanPrice.includes('-')) {
      const parts = cleanPrice.split('-');
      const prices = parts.map(p => parseFloat(p.replace(/,/g, '')));
      return Math.min(...prices.filter(p => !isNaN(p)));
    }
    
    // Handle "From $1,583" cases
    const matches = cleanPrice.match(/(\d+(?:,\d+)*(?:\.\d+)?)/);
    if (matches && matches[1]) {
      return parseFloat(matches[1].replace(/,/g, ''));
    }
    
    return 0;
  }

  /**
   * Parse square footage from strings
   */
  static parseSquareFootage(sqftString: string): number {
    if (!sqftString) return 0;
    
    const matches = sqftString.match(/(\d+(?:,\d+)*)/);
    if (matches && matches[1]) {
      return parseInt(matches[1].replace(/,/g, ''));
    }
    
    return 0;
  }

  /**
   * Parse bedroom count from strings
   */
  static parseBedrooms(bedroomString: string): number {
    if (!bedroomString) return 0;
    
    // Handle "Studio" cases
    if (bedroomString.toLowerCase().includes('studio')) return 0;
    
    const matches = bedroomString.match(/(\d+)\s*(beds?|bedrooms?|br)/i);
    if (matches && matches[1]) {
      return parseInt(matches[1]);
    }
    
    return 0;
  }

  /**
   * Parse bathroom count from strings (handles .5 bathrooms)
   */
  static parseBathrooms(bathroomString: string): number {
    if (!bathroomString) return 0;
    
    // Look for decimal bathrooms first (e.g., "2.5")
    const decimalMatch = bathroomString.match(/(\d+\.5)\s*(baths?|bathrooms?|ba)/i);
    if (decimalMatch && decimalMatch[1]) {
      return parseFloat(decimalMatch[1]);
    }
    
    // Look for whole number bathrooms
    const wholeMatch = bathroomString.match(/(\d+)\s*(baths?|bathrooms?|ba)/i);
    if (wholeMatch && wholeMatch[1]) {
      return parseInt(wholeMatch[1]);
    }
    
    return 0;
  }

  /**
   * Clean and format floor plan names
   */
  static cleanFloorPlanName(name: string): string {
    if (!name) return 'Unknown Plan';
    
    // Remove extra whitespace and newlines
    let cleaned = name.replace(/\s+/g, ' ').trim();
    
    // Remove redundant bedroom/bathroom info if it's already parsed separately
    cleaned = cleaned.replace(/\d+\s*(bed|bedroom|bath|bathroom|br|ba)\s*/gi, '');
    cleaned = cleaned.replace(/\d+,?\d*\s*sq\.?\s*ft\.?/gi, '');
    cleaned = cleaned.replace(/from\s*\$[\d,]+/gi, '');
    cleaned = cleaned.replace(/view\s*\d*\s*apartments?/gi, '');
    
    // Capitalize properly
    cleaned = cleaned.split(' ')
      .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join(' ');
    
    return cleaned.trim() || 'Unknown Plan';
  }

  /**
   * Determine unit type based on name and features
   */
  static determineUnitType(name: string, bathrooms: number): 'apartment' | 'townhome' | 'studio' {
    const lowerName = name.toLowerCase();
    
    if (lowerName.includes('studio')) return 'studio';
    if (lowerName.includes('townhome') || lowerName.includes('townhouse')) return 'townhome';
    if (bathrooms > 2.5) return 'townhome'; // Likely a townhome if it has 3+ bathrooms
    
    return 'apartment';
  }

  /**
   * Calculate price per square foot
   */
  static calculatePricePerSqFt(price: number, sqft: number): number {
    if (price <= 0 || sqft <= 0) return 0;
    return Math.round((price / sqft) * 100) / 100; // Round to 2 decimal places
  }

  /**
   * Standardize availability information
   */
  static standardizeAvailability(availability: string): string {
    if (!availability) return 'Contact for Availability';
    
    const lower = availability.toLowerCase();
    
    if (lower.includes('available now') || lower === 'available') {
      return 'Available Now';
    }
    
    if (lower.includes('not specified') || lower.includes('not available')) {
      return 'Contact for Availability';
    }
    
    // Try to parse dates
    const dateMatch = availability.match(/(\d{1,2}\/\d{1,2}\/\d{4})/);
    if (dateMatch) {
      return `Available ${dateMatch[1]}`;
    }
    
    return availability;
  }

  /**
   * Create property summary from floor plans
   */
  static createPropertySummary(propertyName: string, floorPlans: FloorPlan[]): PropertySummary {
    const prices = floorPlans.map(fp => fp.price).filter(p => p > 0);
    const pricesPerSqFt = floorPlans.map(fp => fp.pricePerSqFt).filter(p => p > 0);
    
    const bedroomCounts = floorPlans.reduce((acc, fp) => {
      switch (fp.bedrooms) {
        case 0: acc.studio++; break;
        case 1: acc.oneBed++; break;
        case 2: acc.twoBed++; break;
        case 3: acc.threeBed++; break;
        default: acc.fourPlusBed++; break;
      }
      return acc;
    }, { studio: 0, oneBed: 0, twoBed: 0, threeBed: 0, fourPlusBed: 0 });

    return {
      propertyName,
      totalFloorPlans: floorPlans.length,
      priceRange: {
        min: Math.min(...prices),
        max: Math.max(...prices)
      },
      avgPricePerSqFt: Math.round((pricesPerSqFt.reduce((a, b) => a + b, 0) / pricesPerSqFt.length) * 100) / 100,
      bedroomDistribution: bedroomCounts,
      availableUnits: floorPlans.filter(fp => 
        fp.availability.includes('Available') || fp.availability.includes('/')
      ).length
    };
  }

  /**
   * Create comprehensive daily report
   */
  static createDailyReport(propertySummaries: PropertySummary[], allFloorPlans: FloorPlan[]): DailyReport {
    const prices = allFloorPlans.map(fp => fp.price).filter(p => p > 0);
    const pricesPerSqFt = allFloorPlans.map(fp => fp.pricePerSqFt).filter(p => p > 0);
    
    const cheapest = allFloorPlans.reduce((min, fp) => 
      fp.price > 0 && (min.price === 0 || fp.price < min.price) ? fp : min
    );
    
    const mostExpensive = allFloorPlans.reduce((max, fp) => 
      fp.price > max.price ? fp : max
    );
    
    const bestValue = allFloorPlans.reduce((best, fp) => 
      fp.pricePerSqFt > 0 && (best.pricePerSqFt === 0 || fp.pricePerSqFt < best.pricePerSqFt) ? fp : best
    );

    return {
      date: new Date().toISOString().split('T')[0] || new Date().toDateString(),
      properties: propertySummaries,
      allFloorPlans,
      marketSummary: {
        totalUnits: allFloorPlans.length,
        avgRent: Math.round(prices.reduce((a, b) => a + b, 0) / prices.length),
        avgPricePerSqFt: Math.round((pricesPerSqFt.reduce((a, b) => a + b, 0) / pricesPerSqFt.length) * 100) / 100,
        cheapestUnit: cheapest,
        mostExpensiveUnit: mostExpensive,
        bestValueUnit: bestValue
      }
    };
  }
}
